modules/core/src/main/java/org/apache/ignite/internal/util/nio/GridNioServer.java

```diff
@@ -4507,4 +4507,16 @@ interface SessionChangeRequest {
          */
         NioOperation operation();
     }
+
+    private static @Nullable GridSelectorNioSessionImpl sessionForCancelConnect(SelectionKey key) {
+        Object a = key.attachment();
+
+        if (a instanceof GridSelectorNioSessionImpl)
+            return (GridSelectorNioSessionImpl) a;
+        else if (a instanceof NioOperationFuture && ((NioOperationFuture<?>) a).operation() == NioOperation.CONNECT)
+            // Canceled immediately after NioOperation.CONNECT without waiting NioOperation.REGISTER.
+            return null;
+
+        throw new IgniteException("No session found when closing connection");
```

Comment:
Should we simply return `null` as well? There's this code after, it'll handle everything I suppose
```
if (ses == null) {
    // Session was not created yet.
    if (key != null)
        key.cancel();

    U.closeQuiet(ch);
}
```